### ISSUE TITLE
TP10_EJ6

### ISSUE THREAD

**Issue by dalilaorbaj** [PI-ITBA/2024_01#222]
Hola! Tengo una duda sobre mi resolucion del ej 6: 

```c
void deleteDupl(TList l){
    if(l==NULL || l->tail == NULL){
        return ;
    }
    if(l->elem == l->tail->elem){
        l->tail = l->tail->tail;
        /*
        La solucion reemplaza mi linea de arriba por lo siguientt: 
            TList aux = list->tail;
            list->tail = aux->tail;
            free(aux);
        **(?) Es necesario? Es de mejor practica? O hay algun otro beneficio?**
        */
        deleteDupl(l);
        return ;
    }
    deleteDupl(l->tail);
}
```

**Comment by ImNotGone** [comment]
Si es necesario hacer el free preguntas?
En caso de no hacer el free tendrias leaks de memoria, no deberias tenerlos nunca. Tener leaks de memoria lleva a muchos inconvenientes, sobretodo cuando  el programa es de ejecucion prolongada, puede llevar a dejar la computadora sin memoria.

**Comment by dalilaorbaj** [comment]
No, a lo que me referia es si esta bien reemplazar estas lineas: 
```
TList aux = list->tail;
list->tail = aux->tail;
free(aux);
```

Por la siguiente linea: 
```c
l->tail = l->tail->tail;
```

**Comment by marcelogarberoglio** [comment]
No, no estaría bien. 
Como bien dijo @ImNotGone no estarías liberando el nodo. Si bien "funciona" porque la lista queda consistente, no se considera que esté bien si hay memoria reservada que no se libera. Esta podría ser una función auxiliar de un programa que maneja miles de elementos, y hacerlo así podría provocar que te quedes sin memoria.

**Comment by dalilaorbaj** [comment]
pero no se lo pisaria con la informacion de l->tail->tail?

**Comment by marcelogarberoglio** [comment]
Sí, estarías eliminando de la lista el nodo siguiente. Ese nodo siguiente ya no se puede acceder, es como si no estuviera. Pero sigue ocupando memoria en el heap, memoria que ya no se podrá usar en otros mallocs o similares. Si tu programa además de borrar nodos después tiene que agregar no va a poder aprovechar esa memoria en desuso, capaz tenga que pedirle al sistema operativo que agrande la memoria para el stack, incluso desplazarla a otro bloque de la RAM, en resumen: uso ineficiente de recursos.

Por eso les pedimos que compilen con el flag de sanitizer. Vas a ver que al final imprime "OK" porque pasó los asserts, pero sanitize te va informar que quedó memoria en el heap que no liberaste.

**Comment by ImNotGone** [comment]
Suponete que tenes los siguientes nodos en las direcciones respectivas.

\- | N0 | N1| N2
-------|----|----|----
dir memoria |  0xA0 | 0xB0 | 0xC0
elem | 1 | 2 | 3
tail    |0xB0|0xC0|NULL

Si eliminas el N1 de tu forma, estarias simplemente reescibiendo el tail de N0
entonces quedaria asi:

\- | N0 | N1| N2
-------|----|----|----
dir memoria |  0xA0 | 0xB0 | 0xC0
elem | 1 | 2 | 3
tail    |`0xC0`|0xC0|NULL

N1 seguiria existindo y ahora perdiste el unico puntero que te indicaba donde esta. Terminan siendo 8 bytes de la ram que ahora figuran como ocupados pero tu programa ya no los esta usando. Ahora estas manejando pocos nodos, pero si tu programa no libera la memoria que usa eventualmente todos los bytes que vas dejando colgados te llenan la ram

**Comment by dalilaorbaj** [comment]
Ahhh okey ahi entendi, muchisimas gracias!!!
